File: go/pkg/genx/profilers/mux.go
package profilers

import (
	"context"
	"fmt"

	"github.com/haivivi/giztoy/go/pkg/trie"
)

// DefaultMux is the default profiler multiplexer.
var DefaultMux = NewMux()

// Handle registers a profiler for the given pattern to the default mux.
func Handle(pattern string, p Profiler) error {
	return DefaultMux.Handle(pattern, p)
}

// Get returns the profiler registered for the given pattern from the default mux.
func Get(pattern string) (Profiler, error) {
	return DefaultMux.Get(pattern)
}

// Process runs the profiler registered for the given pattern in the default mux.
func Process(ctx context.Context, pattern string, input Input) (*Result, error) {
	return DefaultMux.Process(ctx, pattern, input)
}

// Mux is a profiler multiplexer that routes processing requests to registered
// [Profiler] implementations based on pattern matching using a trie.
type Mux struct {
	mux *trie.Trie[Profiler]
}

// NewMux creates a new profiler multiplexer.
func NewMux() *Mux {
	return &Mux{
		mux: trie.New[Profiler](),
	}
}

// Handle registers a profiler for the given pattern.
// Returns an error if a profiler is already registered for the pattern.
func (m *Mux) Handle(pattern string, p Profiler) error {
	return m.mux.Set(pattern, func(ptr *Profiler, existed bool) error {
		if existed {
			return fmt.Errorf("profilers: profiler already registered for %s", pattern)
		}
		*ptr = p
		return nil
	})
}

// Get returns the profiler registered for the given pattern.
func (m *Mux) Get(pattern string) (Profiler, error) {
	ptr, ok := m.mux.Get(pattern)
	if !ok {
		return nil, fmt.Errorf("profilers: profiler not found for %s", pattern)
	}
	p := *ptr
	if p == nil {
		return nil, fmt.Errorf("profilers: profiler not found for %s", pattern)
	}
	return p, nil
}

// Process runs the profiler registered for the given pattern.
func (m *Mux) Process(ctx context.Context, pattern string, input Input) (*Result, error) {
	p, err := m.Get(pattern)
	if err != nil {
		return nil, err
	}
	return p.Process(ctx, input)
}

File: rust/genx/src/profilers/mux.rs
//! Profiler multiplexer.

use std::collections::HashMap;
use std::sync::Arc;

use crate::error::GenxError;

use super::types::{Profiler, ProfilerInput, ProfilerResult};

/// Profiler multiplexer that routes requests to registered implementations.
pub struct ProfilerMux {
    routes: HashMap<String, Arc<dyn Profiler>>,
}

impl ProfilerMux {
    pub fn new() -> Self {
        Self {
            routes: HashMap::new(),
        }
    }

    pub fn handle(
        &mut self,
        pattern: impl Into<String>,
        p: Arc<dyn Profiler>,
    ) -> Result<(), GenxError> {
        let pattern = pattern.into();
        if self.routes.contains_key(&pattern) {
            return Err(GenxError::Other(anyhow::anyhow!(
                "profilers: profiler already registered for {}",
                pattern,
            )));
        }
        self.routes.insert(pattern, p);
        Ok(())
    }

    pub fn get(&self, pattern: &str) -> Result<&dyn Profiler, GenxError> {
        self.routes
            .get(pattern)
            .map(|p| p.as_ref())
            .ok_or_else(|| {
                GenxError::Other(anyhow::anyhow!(
                    "profilers: profiler not found for {}",
                    pattern,
                ))
            })
    }

    pub async fn process(
        &self,
        pattern: &str,
        input: ProfilerInput,
    ) -> Result<ProfilerResult, GenxError> {
        let p = self.get(pattern)?;
        p.process(input).await
    }
}

impl Default for ProfilerMux {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use async_trait::async_trait;

    struct FakeProfiler {
        name: String,
    }

    #[async_trait]
    impl Profiler for FakeProfiler {
        async fn process(&self, _input: ProfilerInput) -> Result<ProfilerResult, GenxError> {
            Ok(ProfilerResult {
                schema_changes: vec![],
                profile_updates: HashMap::new(),
                relations: vec![],
            })
        }
        fn model(&self) -> &str {
            &self.name
        }
    }

    #[tokio::test]
    async fn t12_5_mux_register_and_route() {
        let mut mux = ProfilerMux::new();
        mux.handle("test", Arc::new(FakeProfiler { name: "test".into() }))
            .unwrap();

        use crate::segmentors::{SegmentOutput, SegmentorResult};
        let input = ProfilerInput {
            messages: vec![],
            extracted: SegmentorResult {
                segment: SegmentOutput {
                    summary: "".into(),
                    keywords: vec![],
                    labels: vec![],
                },
                entities: vec![],
                relations: vec![],
            },
            schema: None,
            profiles: None,
        };
        let result = mux.process("test", input).await.unwrap();
        assert!(result.schema_changes.is_empty());
    }

    #[tokio::test]
    async fn t12_5_mux_not_found() {
        let mux = ProfilerMux::new();
        use crate::segmentors::{SegmentOutput, SegmentorResult};
        let input = ProfilerInput {
            messages: vec![],
            extracted: SegmentorResult {
                segment: SegmentOutput {
                    summary: "".into(),
                    keywords: vec![],
                    labels: vec![],
                },
                entities: vec![],
                relations: vec![],
            },
            schema: None,
            profiles: None,
        };
        assert!(mux.process("missing", input).await.is_err());
    }
}
